### PR TITLE
[llvm][LoongArch] Support per-global code model attribute for LoongArch

### PR DESCRIPTION
**Pull request by heiher**
This patch gets the code model from global variable attribute if it has, otherwise the target's will be used.

### CODE REVIEW

**Comment by heiher** [comment]
Part 1: #72077 
Part 2: #72078

**Comment by heiher** [comment]
cc @xen0n @xry111

**Comment by xen0n** [comment]
I think we should initially only support the attribute on LoongArch. GCC's `model` attribute is only ever supported on ia64 and loongarch, yet the patch currently seems to allow other architectures to use this attribute, with LLVM code model names accepted -- this may or may not be what the targets actually want, if they decide to add support for the attribute in GCC.

**Comment by heiher** [comment]
> I think we should initially only support the attribute on LoongArch. GCC's `model` attribute is only ever supported on ia64 and loongarch, yet the patch currently seems to allow other architectures to use this attribute, with LLVM code model names accepted -- this may or may not be what the targets actually want, if they decide to add support for the attribute in GCC.

Done. Thanks

File: llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp [hunk 1]
```diff
@@ -766,24 +767,37 @@ SDValue LoongArchTargetLowering::getAddr(NodeTy *N, SelectionDAG &DAG,
...
 
 SDValue LoongArchTargetLowering::lowerGlobalAddress(SDValue Op,
                                                     SelectionDAG &DAG) const {
   GlobalAddressSDNode *N = cast<GlobalAddressSDNode>(Op);
   assert(N->getOffset() == 0 && "unexpected offset in global node");
-  return getAddr(N, DAG, N->getGlobal()->isDSOLocal());
+  auto CM = DAG.getTarget().getCodeModel();
+  const GlobalValue *GV = N->getGlobal();
+
+  if (GV->isDSOLocal()) {
```
Comment (SixWeining) [inline llvm/lib/Target/LoongArch/LoongArchISelLowering.cpp:793]: Can we merge these 3 `if`s into one?

**Review by SixWeining — APPROVED** [review]
LGTM with a small nit.

**Pull request merged**